rotkehlchen/chain/solana/manager.py

```diff
@@ -0,0 +1,95 @@
+from rotkehlchen.chain.evm.types import WeightedNode
+from rotkehlchen.chain.mixins.rpc_nodes import SolanaRPCMixin
+from rotkehlchen.chain.solana.utils import lamports_to_sol
+from rotkehlchen.constants.misc import ZERO
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.fval import FVal
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import SolanaAddress, SupportedBlockchain
+
+if TYPE_CHECKING:
+    from rotkehlchen.db.dbhandler import DBHandler
+    from rotkehlchen.greenlets.manager import GreenletManager
+
+R = TypeVar('R')
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class SolanaManager(SolanaRPCMixin):
+
+    def __init__(
+            self,
+            greenlet_manager: 'GreenletManager',
+            database: 'DBHandler',
+    ):
+        SolanaRPCMixin.__init__(self)
+        self.greenlet_manager = greenlet_manager
+        self.database = database
+        self.blockchain = SupportedBlockchain.SOLANA
+        self.rpc_timeout = DEFAULT_RPC_TIMEOUT
+
+    def _query(
+            self,
+            method: Callable[[Client], R],
+            call_order: Sequence[WeightedNode],
+    ) -> R:
+        """Use the provided call order to request fn from solana"""
+        for weghted_node in call_order:
+            node_info = weghted_node.node_info
+            if (rpc_node := self.rpc_mapping.get(node_info, None)) is None:
+                if node_info.name in self.failed_to_connect_nodes:
+                    continue
+
+                success, _ = self.attempt_connect(node=node_info)
+                if success is False:
+                    self.failed_to_connect_nodes.add(node_info.name)
+                    continue
+
+                if (rpc_node := self.rpc_mapping.get(node_info, None)) is None:
+                    log.error(f'Unexpected missing node {node_info} at Solana')
+                    continue
+
+            try:
+                return method(rpc_node.rpc_client)
+            except (HTTPError, RPCException) as e:
+                log.error(f'Failed to call {node_info.name} due to {e}')
+
+        log.error(f'Tried all solana nodes in {call_order} for {method} but did not get any response')  # noqa: E501
+        raise RemoteError(f'Failed to get {method}')
+
+    def get_multi_balance(self, accounts: Sequence[SolanaAddress]) -> dict[SolanaAddress, FVal]:
+        """Returns a dict with keys being accounts and balances in the chain native token.
+
+        In case of remote error it returns ZERO for all the accounts.
+        """
+        try:
+            response = self._query(
+                method=lambda client: client.get_multiple_accounts(
+                    pubkeys=[Pubkey.from_string(addr) for addr in accounts],
+                ),
+                call_order=self.default_call_order(),
+            )
+        except RemoteError as e:
+            log.error(f'Failed to query solana balances for {accounts} due to {e}')
```

Comment:
Shouldn't this be a failure and retry from another node? Not just silently return zero balance?

Reply:
query retries all the available nodes and raises when the nodes are exhausted and all returned an error. `default_call_order` returns the list of nodes to call